ouroboros-consensus/src/Ouroboros/Consensus/Protocol/PBFT/ChainState.hs

```diff
@@ -128,6 +130,10 @@ data PBftChainState c = PBftChainState {
 
       -- | Cached counts of the signatures in the window
     , counts     :: !(Map (PBftVerKeyHash c) Word64)
+
+      -- | Map from slots known to contain epoch boundary blocks (EBBs) to the
+      -- latest preceding slot known to contain a non-EBB block
+    , ebbs       :: !(Map SlotNo (WithOrigin SlotNo))
```

Comment:
Maybe mention that EBBs are not signed and don't contribute to the counts.

Which invariants could we formulate for `ebbs`?

* For each `(k, v)` in `ebbs`, `v <= At k`: maps to *preceding* slots
* What is the relation to `pre/inWindow`?
* What is the relation to `pre/postAnchor`?

Reply:
I pushed a commit that ensures (still no tests yet) and documents the following invariants. It also mentions in a few comments that EBBs are unsigned -- that is an excellent clarifying observation!

```
      -- INVARIANT For all @(ebbSlot, mSlot)@ in @'ebbs' (cs :: 'ChainState')@,
      --
      --  * @mSlot >= anchorSlot cs@; see 'pruneEBBsLT'
      --
      --  * @'At' ebbSlot <= tgt@ if @cs@ is the result of a 'rewind' to @tgt@;
      --    see 'pruneEBBsGT'
      --
      --  * @and [ At s <= mSlot | s <- precedingSignedSlots ]@
      --
      --  * @'rewind' k n ('At' ebbSlot) cs = 'rewind' k n mSlot cs@
      --
      --   where
      --
      --  * @precedingSignedSlots = Seq.filter (< ebbSlot) signedSlots@
      --
      --  * @signedSlots = 'pbftSignerSlotNo' <$> ('preAnchor' <> 'postAnchor')@
```

and

```
-- INVARIANT Key greater: For all @(k, v)@, @At k > v@.
--
-- INVARIANT Non-descending: For all @(k1, v1)@ and @(k2, v2)@, @k1 < k2@
-- implies @v1 <= v2@.
```

Reply:
These invariants are now included in `assertInvariant`.

(Also, previous commits added EBBs to the the existing PBFT ChainState tests.)